src/Parsers/ASTIdentifier.cpp

```diff
@@ -16,59 +16,65 @@ namespace ErrorCodes
     extern const int SYNTAX_ERROR;
 }
 
+ASTIdentifier::ASTIdentifier(const String & short_name)
+    : full_name(short_name), name_parts{short_name}, semantic(std::make_shared<IdentifierSemanticImpl>())
+{
+    assert(!full_name.empty());
+}
 
-ASTPtr ASTIdentifier::clone() const
+ASTIdentifier::ASTIdentifier(std::vector<String> && name_parts_, bool special)
+    : name_parts(name_parts_), semantic(std::make_shared<IdentifierSemanticImpl>())
 {
-    auto ret = std::make_shared<ASTIdentifier>(*this);
-    ret->semantic = std::make_shared<IdentifierSemanticImpl>(*ret->semantic);
-    return ret;
+    assert(!name_parts.empty());
+    for (const auto & part : name_parts)
+    {
+        assert(!part.empty());
+        (void) part;  // otherwise not-used in release build
+    }
+
+    semantic->special = special;
+    semantic->legacy_compound = true;
+
+    if (!special && name_parts.size() >= 2)
+        semantic->table = name_parts.end()[-2];
 }
 
-std::shared_ptr<ASTIdentifier> ASTIdentifier::createSpecial(const String & name, std::vector<String> && name_parts)
+ASTPtr ASTIdentifier::clone() const
 {
-    auto ret = std::make_shared<ASTIdentifier>(name, std::move(name_parts));
-    ret->semantic->special = true;
+    auto ret = std::make_shared<ASTIdentifier>(*this);
+    ret->semantic = std::make_shared<IdentifierSemanticImpl>(*ret->semantic);
     return ret;
 }
 
-ASTIdentifier::ASTIdentifier(const String & name_, std::vector<String> && name_parts_)
-    : name(name_)
-    , name_parts(name_parts_)
-    , semantic(std::make_shared<IdentifierSemanticImpl>())
+bool ASTIdentifier::supposedToBeCompound() const
 {
-    if (!name_parts.empty() && name_parts[0].empty())
-        name_parts.erase(name_parts.begin());
-
-    if (name.empty())
-    {
-        if (name_parts.size() == 2)
-            name = name_parts[0] + '.' + name_parts[1];
-        else if (name_parts.size() == 1)
-            name = name_parts[0];
-    }
+    return semantic->legacy_compound;
 }
 
-ASTIdentifier::ASTIdentifier(std::vector<String> && name_parts_)
-    : ASTIdentifier("", std::move(name_parts_))
-{}
-
 void ASTIdentifier::setShortName(const String & new_name)
 {
-    name = new_name;
-    name_parts.clear();
+    assert(!new_name.empty());
+
+    full_name = new_name;
+    name_parts = {new_name};
 
     bool special = semantic->special;
     *semantic = IdentifierSemanticImpl();
     semantic->special = special;
 }
 
-void ASTIdentifier::restoreCompoundName()
+const String & ASTIdentifier::name() const
 {
-    if (name_parts.empty())
-        return;
-    name = name_parts[0];
-    for (size_t i = 1; i < name_parts.size(); ++i)
-        name += '.' + name_parts[i];
+    assert(!name_parts.empty());
+
+    if (full_name.empty())
+    {
+        full_name = name_parts[0];
```

Comment:
Do I understand correctly, that it cannot be called from multiple threads (or, at least, `full_name` is initialized on early stages of query processing when there is only one thread)?

Reply:
It's a good point to consider - I will recheck the name initialization